### PR TITLE
Rubyfmt integration

### PR DESCRIPTION
**Pull request by penelope-stripe**
adds a bazel build for rubyfmt. There's plenty I'd like some help on here:

1. the skylark changes I made to expose CcInfo seems hacky, I'm not sure if there's a better way to do it.
2. Rubyfmt's source had to be mangled a little to get this working, I'll need to provide some scripts to make it easier to update.
3. I had to remove jemalloc from rubyfmt. This impacts performance a little bit. It'd be nice to work out why the linking symbols were removed from `librubyfmt` during the final compilation.

I am also horrible at bazel, any style review on that would be lovely.

### Motivation
This is so we can integrate rubyfmt in to the sorbet lsp

### Test plan
./bazel run -s //experimental/rubyfmt_example_c_invocation:main

### CODE REVIEW

File: core/Context.cc
Comment (jvilk-stripe) [inline core/Context.cc:80]: Hmm, did you merge in changes from master? This is a diff from a change I recently merged to master.

**Review by jvilk-stripe — CHANGES_REQUESTED** [review]
Thanks for looking into this and taking this on. I'm excited to let people try Rubyfmt integration!

I didn't realize that you were working on adding all of Rubyfmt's source code to the Sorbet tree! I was expecting that Rubyfmt would be another third party dependency that we download and build. Is it possible to make the required bazel changes upstream? Right now, this PR significantly expands the source tree of Sorbet, so I cannot accept it. :( 

If not, then maybe we should go back to the drawing board and invoke rubyfmt as an external process. It's not as convenient as a simple library API, but it should do for an experimental feature. If we end up going this route, I'd like to apologize if this Bazel work ended up taking a significant amount of time. :(

File: core/Context.cc
Comment (penelope-stripe) [reply core/Context.cc:80]: this looks like a bad rebase. I will fix!

**Comment by penelope-stripe** [comment]
@jvilk-stripe PTAL. This should be ready to go now.

**Comment by penelope-stripe** [comment]
Notes:

* this adds a script `tools/scripts/cargo_raze.sh` which runs `cargo raze` to regenerate the cargo build files. This will require `https://github.com/google/cargo-raze` to be installed. That is, it is not managed by bazel.
* you can try echo 'puts "hi"' | ./bazel run //experimental/rubyfmt_c_main:main to see this working
* this currently does a debug build which does a lot of logging. We can do a release build if you would like
* compiling with jemalloc proved to be ~close to impossible, so I'm using Rust's built in malloc. I believe this will use jemalloc if the entire binary is compiled with jemalloc, but I'm not sure how to prove that.

**Comment by jvilk-stripe** [comment]
Hmm, the build failed on my Linux devbox :( :

```
ERROR: /pay/src/sorbet/experimental/rubyfmt_c_main/BUILD:1:10: Linking of rule '//experimental/rubyfmt_c_main:main' failed (Exit 1)
ld.lld: error: undefined symbol: timer_settime
>>> referenced by thread.c
>>>               thread.o:(ubf_timer_arm) in archive bazel-out/k8-dbg/bin/external/ruby_2_6/toolchain/static_libs/libruby-static.a
```

**Comment by jvilk-stripe** [comment]
This now builds on my devbox. I'm going to see if we can build it in Stripe's CI system.

File: experimental/rubyfmt_c_main/BUILD
Comment (jez) [inline experimental/rubyfmt_c_main/BUILD:1]: Is this binary here just to test the build? i.e., do we expect to replace this `cc_binary` with a library rule eventually when integrating it into sorbet's LSP?

File: WORKSPACE
Comment (jez) [inline WORKSPACE:93]: Is there a reason why these archives must be in WORKSPACE? Can you move them to `third_party/externals.bzl` where all the rest of the third party dependencies are? (we have third party deps in a separate function to make it easier to depend on sorbet as a bazel library and get the same deps)

File: WORKSPACE
Comment (jez) [inline WORKSPACE:111]: What is this rule for?

File: third_party/ruby/build-ruby.bzl
Comment (jez) [inline third_party/ruby/build-ruby.bzl:491]: cc @elliottt wanted to flag that there were some changes to the build-ruby.bzl script here.

File: third_party/externals.bzl
Comment (jez) [inline third_party/externals.bzl:153]: Would it be possible for you to use an http_archive for this? Stripe's internal CI places limits on which external services we can call out to from behind a proxy, and GitHub is excluded (it's why you see all those artifactory fallback URLs in this file)

File: tools/scripts/cargo_raze.sh
Comment (jez) [inline tools/scripts/cargo_raze.sh:7]: have you documented this script in a README somewhere?

i'd love to have a description of when and how to update rubyfmt's deps etc.

File: third_party/cargo/remote/wasi-0.10.0+wasi-snapshot-preview1.BUILD
Comment (jez) [inline third_party/cargo/remote/wasi-0.10.0+wasi-snapshot-preview1.BUILD:28]: ooc, what do you depend on WASI for?

File: third_party/cargo/Cargo.toml
Comment (jez) [inline third_party/cargo/Cargo.toml:23]: Are these dependencies duplicated out of rubyfmt's source tree? How will we know if these are out of date / how to update them?

**Comment by jez** [comment]
This change looks really good. Glad to see `cargo-raze`; i've had success with that project in the past and it's nice to build on pre-existing tools (i'd love for some sort of bazel support for building / running Ruby for example, rather than the pile of bazel we have in sorbet).

I left some comments & questions. Thanks!

File: experimental/rubyfmt_c_main/BUILD
Comment (penelope-stripe) [reply experimental/rubyfmt_c_main/BUILD:1]: yes, eventually this would just end up as part of Sorbet's build, but this makes sure everything is working.

File: WORKSPACE
Comment (penelope-stripe) [reply WORKSPACE:93]: I was following the guide from io_bazel_rules_rust, will update

File: WORKSPACE
Comment (penelope-stripe) [reply WORKSPACE:111]: image
from rules_rust docs

File: third_party/externals.bzl
Comment (penelope-stripe) [reply third_party/externals.bzl:153]: I can provide this as a http archive.

File: tools/scripts/cargo_raze.sh
Comment (penelope-stripe) [reply tools/scripts/cargo_raze.sh:7]: I'll write a readme for this

File: third_party/ruby/build-ruby.bzl
Comment (elliottt) [inline third_party/ruby/build-ruby.bzl:476]: Could you remove the `print` if it was just used for debugging?

File: third_party/ruby/build-ruby.bzl
Comment (elliottt) [inline third_party/ruby/build-ruby.bzl:237]: What is the `CcInfo` provider used for?

File: third_party/cargo/remote/wasi-0.10.0+wasi-snapshot-preview1.BUILD
Comment (penelope-stripe) [reply third_party/cargo/remote/wasi-0.10.0+wasi-snapshot-preview1.BUILD:28]: this is actually not an intentional dependency, it's being pulled in because raze compiles all child crates with all features enabled, so in the strictest sense is not needed. That said, I don't see an option in raze to disable that, so it'll likely have to be added for now

File: third_party/cargo/Cargo.toml
Comment (penelope-stripe) [reply third_party/cargo/Cargo.toml:23]: They are. This is the most fragile part, so I'll write a readme for this.

File: third_party/ruby/build-ruby.bzl
Comment (elliottt) [inline third_party/ruby/build-ruby.bzl:463]: `find_cpp_toolchain` relies on having the `_cc_toolchain` attribute present on the rule for compatibility, as in the `_build_ruby` rule definition above:
https://github.com/sorbet/sorbet/blob/2f624af4c4b086dde25c06bd78502fc1fbc8a7bf/third_party/ruby/build-ruby.bzl#L253-L255

File: third_party/ruby/build-ruby.bzl
Comment (penelope-stripe) [reply third_party/ruby/build-ruby.bzl:237]: It's because `rubyfmt_static_deps_impl` depends on this. When a `cc_binary` targets `rubyfmt_static_deps`, it expects a `ccinfo` from that, but also from everything `rubyfmt_static_deps` depends on transitively, however, because we don't need anything specifically from this step, it's just a noop

File: third_party/cargo/remote/wasi-0.10.0+wasi-snapshot-preview1.BUILD
Comment (jez) [reply third_party/cargo/remote/wasi-0.10.0+wasi-snapshot-preview1.BUILD:28]: no worries, i just think WASI is cool and was interested to hear if you had found a use case for it.

File: third_party/externals.bzl
Comment (penelope-stripe) [reply third_party/externals.bzl:152]: done.

File: third_party/ruby/build-ruby.bzl
Comment (penelope-stripe) [reply third_party/ruby/build-ruby.bzl:463]: we dealt with this yesterday!

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_I9N5NHFT0GRKs0

File: third_party/cargo/add_crates_shas.rb
Comment (elliottt) [inline third_party/cargo/add_crates_shas.rb:14]: Thanks for automating this!

File: tools/scripts/cargo_raze.sh
Comment (elliottt) [inline tools/scripts/cargo_raze.sh:7]: Is this intermediate call to `format_build_files.sh` needed for the subsequent call to `add_crates_shas.rb`?

**Review by elliottt — APPROVED** [review]
This looks great to me! I only have one question about the `cargo_raze.sh` script, but it's not blocking.

**Comment by elliottt** [comment]
> We have a policy of testing changes to Sorbet against Stripe's codebase before
> merging them. I've kicked off a test run for the current PR. When the build
> finishes, I'll share with you whether or how it failed. Thanks!
> 
> Stripe employees can see the build result here:
> 
> → https://go/builds/bui_I9N5NHFT0GRKs0

This build passed, but I don't think that it built rubyfmt. Let's not merge until we get a build that definitely contains rubyfmt in CI.

**Comment by elliottt** [comment]
@penelope-stripe there was a build error related to BUILD files missing in parts of the rubyfmt config: https://go/builds/bui_I9PrvOazO6jGHB

**Comment by elliottt** [comment]
We can redirect the toolchain download internally using `STATIC_RUST_URL`: https://github.com/bazelbuild/rules_rust/pull/143

**Comment by elliottt** [comment]
I got a successful build here: https://cibot.corp.stripe.com/builds/bui_IEYzh3cv5KTlqU

There's a little bit of integration work to be done, but that will be easy to land separately.

**Review by jvilk-stripe — APPROVED** [review]

**Comment by elliottt** [comment]
Once https://go/ps/pull/275544 is merged, we can merge this in 👍

File: tools/scripts/cargo_raze.sh
Comment (elliottt) [reply tools/scripts/cargo_raze.sh:7]: Resolved via DM: add_crates_shas.rb is sensitive to the formatting of build files, so both calls are needed.

**Pull request merged**
